### DISCUSSION TITLE
get node ids from a sideset and paired nodes in an Action

### DISCUSSION THREAD

**Discussion by KhaledNabilSharafeldin** [idaholab/moose#20379] [Q&A Meshing] [answered]
Hello,
I have a list of sideset names `BoundaryName` and looking to get the node ids for all the nodes on these sidesets.

also, if there is a way to get the "PairedBoundaryMapping" for that node easily within action or constraint systems.

Ultimately, I am trying to get a list of primary and secondary node ids.

thanks in advance!

**Comment by GiudGiud** [comment, marked as answer]
Hello

Use the NodalValueSampler and restrict it to those sidesets with the boundary parameter
https://mooseframework.inl.gov/source/vectorpostprocessors/NodalValueSampler.html
The CSV output will have the list of node ids

Guillaume

**Reply by KhaledNabilSharafeldin** [reply]
is there a cleaner way to code it in the action instead?

**Reply by GiudGiud** [reply]
Oh yes for sure if you are willing to code.

You can loop over all boundaries and select the nodes from there. 
Start from ` mesh->get_boundary_info();` and build the node list
https://libmesh.github.io/doxygen/classlibMesh_1_1BoundaryInfo.html

**Reply by KhaledNabilSharafeldin** [reply]
It looks like what I need is being done in `MeshBase::get_info()`, there's `SidesetInfo` struct that store `node_ids`

is there an easy way to get to that info? any suggestions would be appreciated.

thanks again!

**Reply by loganharbour** [reply]
I believe you're looking for https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1BoundaryInfo.html#a4ab2a9be3fe7eaa1d6f9f5bdc7a9856e

**Reply by KhaledNabilSharafeldin** [reply]
@loganharbour i can see how I can use the suggested function to return each boundary node paired to the sideset id, and then forming the sidesets from that.
but does this support distributed mesh?

I would rather get the info from this, it seems to be more direct, I am just not sure how to get to that info in an Action, and the function doesnt return these info.


> It looks like what I need is being done in `MeshBase::get_info()`, there's `SidesetInfo` struct that store `node_ids`
> 
> is there an easy way to get to that info? any suggestions would be appreciated.
> 
> thanks again!

**Reply by loganharbour** [reply]
Oh heh, I'm the one that wrote that in `MeshBase`. 

> but does this support distributed mesh?

No, building the side list just builds information pertaining to the local entities (it may also include ghosted entities, but don't quote me on that).  Why do you need to know the global information? In the case of distributed mesh, you should really only want to act on local entities.

**Reply by KhaledNabilSharafeldin** [reply]
> Why do you need to know the global information?

I'm trying to get nodal ids to apply a constraint to these nodes. I still will need to find the node pairs for those, but I haven't looked into that yet. ('getPairedBoundaryMapping' seems to be the method used)

**Reply by KhaledNabilSharafeldin** [reply]
ok, so i found a way to get the node ids by creating sidesetsfromnormals, then getting the list of nodelist, with the dof_id_type i can get a pointer that holds the position vector and check for the corresponding sides.

but I still have a problem with creating the 8 edge corners as nodesets, since all of this is inaccessible at append_mesh flag , any idea if there's already this functionality implemented? (a nodeset generator for corner nodes)

```
std::vector<BoundaryName> boundary_names{"xp", "xn","yp", "yn","zp", "zn"};

std::vector<std::vector<dof_id_type>> allLists_from_boundaryNames;
   
for (auto &&name : boundary_names)
    allLists_from_boundaryNames.push_back(_mesh->getNodeList(_mesh->getBoundaryID(name)));
```

and `  auto nptr = *_mesh->nodePtr(n); ` were the functions i needed to build this up, if anyone's interested

**Reply by GiudGiud** [reply]
You can specify the points manually with this:
https://mooseframework.inl.gov/docs/site/source/meshgenerators/ExtraNodesetGenerator.html

**Reply by KhaledNabilSharafeldin** [reply]
unfortunately, the `_mesh->dimensionWidth(dim)` is not accessible at this section as well, i can do it manually, but is there a way to get it otherwise?
so basically I neither have the total dimension of the cube or the corner node ids at that point of the code.

thanks again

**Reply by GiudGiud** [reply]
So when writing the input file, you will not know where the nodes are?

If so, you can have a look at the source code of this mesh generator and move its functionality to an user object / something that gets executed later on.
